Strided view changes preventing use of `shape` in `reshape_view()`

Here is a minimally reproducible example of something that works in the released version of xtensor, but with the master branch is broken. I think that it has to do with the recent changes made to strided view somewhere around here:

https://github.com/QuantStack/xtensor/commit/39723c6730d0c1035282ff09802f4a11ec5b4333#diff-22db96c2dbc74ce485107a4a4a64295aL649

```cpp
// [[Rcpp::depends(xtensor)]]
// [[Rcpp::plugins(cpp14)]]

#include <xtensor/xarray.hpp>

void reshape_test() {

  xt::xarray<int> x = {{1, 2, 3}, {4, 5, 6}};

  const std::vector<std::size_t>& new_shape = {3, 2};

  auto x_reshaped = xt::reshape_view(x, new_shape);

}
```

```cpp
In file included from test.cpp:4:
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/xtensor/include/xtensor/xarray.hpp:19:
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/xtensor/include/xtensor/xcontainer.hpp:23:
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/xtensor/include/xtensor/xmath.hpp:26:
In file included from /Library/Frameworks/R.framework/Versions/3.5/Resources/library/xtensor/include/xtensor/xoperation.hpp:21:
/Library/Frameworks/R.framework/Versions/3.5/Resources/library/xtensor/include/xtensor/xstrided_view.hpp:605:16: error: no matching constructor for initialization of 'view_type' (aka 'xstrided_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> >, std::decay_t<xarray_container<uvector<int, allocator<int> >, xt::layout_type::row_major, svector<unsigned long, 4, allocator<unsigned long>, true>, xtensor_expression_tag> &>::static_layout, flat_adaptor_getter<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &> >')
        return view_type(std::forward<E>(e), std::forward<S>(shape), std::move(strides), 0, e.layout());
               ^         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test.cpp:12:25: note: in instantiation of function template specialization 'xt::reshape_view<xt::xarray_container<xt::uvector<int, std::__1::allocator<int> >, xt::layout_type::row_major, xt::svector<unsigned long, 4, std::__1::allocator<unsigned long>, true>, xt::xtensor_expression_tag> &, const std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > &>' requested here
  auto x_reshaped = xt::reshape_view(x, new_shape);
                        ^
/Library/Frameworks/R.framework/Versions/3.5/Resources/library/xtensor/include/xtensor/xstrided_view.hpp:161:9: note: candidate constructor template not viable: 2nd argument ('const std::__1::vector<unsigned long, std::__1::allocator<unsigned long> >') would lose const qualifier
        xstrided_view(CTA&& e, S&& shape, strides_type&& strides, std::size_t offset, layout_type layout) noexcept;
        ^
/Library/Frameworks/R.framework/Versions/3.5/Resources/library/xtensor/include/xtensor/xstrided_view.hpp:163:9: note: candidate constructor not viable: requires single argument 'rhs', but 5 arguments were provided
        xstrided_view(const xstrided_view& rhs) = default;
        ^
1 error generated.
make: *** [test.o] Error 1
```

Even passing in a non-const version of `new_shape` doesn't seem to help

--- a/include/xtensor/xdynamic_view.hpp
+++ b/include/xtensor/xdynamic_view.hpp
@@ -143,8 +143,8 @@ namespace xt
         using slice_type = xtl::variant<detail::xfake_slice<strides_vt>, xkeep_slice<strides_vt>, xdrop_slice<strides_vt>>;
         using slice_vector_type = std::vector<slice_type>;
 
-        template <class CTA>
-        xdynamic_view(CTA&& e, S&& shape, get_strides_t<S>&& strides, std::size_t offset, layout_type layout,
+        template <class CTA, class SA>
+        xdynamic_view(CTA&& e, SA&& shape, get_strides_t<S>&& strides, std::size_t offset, layout_type layout,
                       slice_vector_type&& slices, get_strides_t<S>&& adj_strides) noexcept;
 
         template <class E>
@@ -364,11 +364,11 @@ namespace xt
      ********************************/
 
     template <class CT, class S, layout_type L, class FST>
-    template <class CTA>
-    inline xdynamic_view<CT, S, L, FST>::xdynamic_view(CTA&& e, S&& shape, get_strides_t<S>&& strides,
+    template <class CTA, class SA>
+    inline xdynamic_view<CT, S, L, FST>::xdynamic_view(CTA&& e, SA&& shape, get_strides_t<S>&& strides,
                                                        std::size_t offset, layout_type layout,
                                                        slice_vector_type&& slices, get_strides_t<S>&& adj_strides) noexcept
-        : base_type(std::forward<CTA>(e), std::move(shape), std::move(strides), offset, layout),
+        : base_type(std::forward<CTA>(e), std::forward<SA>(shape), std::move(strides), offset, layout),
           m_slices(std::move(slices)), m_adj_strides(std::move(adj_strides))
     {
     }

--- a/include/xtensor/xstrided_view.hpp
+++ b/include/xtensor/xstrided_view.hpp
@@ -157,8 +157,8 @@ namespace xt
 
         using simd_value_type = xt_simd::simd_type<value_type>;
 
-        template <class CTA>
-        xstrided_view(CTA&& e, S&& shape, strides_type&& strides, std::size_t offset, layout_type layout) noexcept;
+        template <class CTA, class SA>
+        xstrided_view(CTA&& e, SA&& shape, strides_type&& strides, std::size_t offset, layout_type layout) noexcept;
 
         xstrided_view(const xstrided_view& rhs) = default;
         xstrided_view& operator=(const xstrided_view& rhs);
@@ -306,9 +306,9 @@ namespace xt
      * @param layout the layout of the view
      */
     template <class CT, class S, layout_type L, class FST>
-    template <class CTA>
-    inline xstrided_view<CT, S, L, FST>::xstrided_view(CTA&& e, S&& shape, strides_type&& strides, std::size_t offset, layout_type layout) noexcept
-        : base_type(std::forward<CTA>(e), std::move(shape), std::move(strides), offset, layout)
+    template <class CTA, class SA>
+    inline xstrided_view<CT, S, L, FST>::xstrided_view(CTA&& e, SA&& shape, strides_type&& strides, std::size_t offset, layout_type layout) noexcept
+        : base_type(std::forward<CTA>(e), std::forward<SA>(shape), std::move(strides), offset, layout)
     {
     }
     //@}

--- a/include/xtensor/xstrided_view_base.hpp
+++ b/include/xtensor/xstrided_view_base.hpp
@@ -125,8 +125,8 @@ namespace xt
         static constexpr layout_type static_layout = inner_types::layout;
         static constexpr bool contiguous_layout = static_layout != layout_type::dynamic && xexpression_type::contiguous_layout;
 
-        template <class CTA>
-        xstrided_view_base(CTA&& e, undecay_shape&& shape, strides_type&& strides, size_type offset, layout_type layout) noexcept;
+        template <class CTA, class SA>
+        xstrided_view_base(CTA&& e, SA&& shape, strides_type&& strides, size_type offset, layout_type layout) noexcept;
 
         xstrided_view_base(xstrided_view_base&& rhs);
 
@@ -298,12 +298,12 @@ namespace xt
      * @param layout the layout of the view
      */
     template <class D>
-    template <class CTA>
-    inline xstrided_view_base<D>::xstrided_view_base(CTA&& e, undecay_shape&& shape, strides_type&& strides, size_type offset, layout_type layout) noexcept
+    template <class CTA, class SA>
+    inline xstrided_view_base<D>::xstrided_view_base(CTA&& e, SA&& shape, strides_type&& strides, size_type offset, layout_type layout) noexcept
         : m_e(std::forward<CTA>(e)),
           //m_storage(detail::get_flat_storage<undecay_expression>(m_e)),
           m_storage(storage_getter::get_flat_storage(m_e)),
-          m_shape(std::move(shape)),
+          m_shape(std::forward<SA>(shape)),
           m_strides(std::move(strides)),
           m_offset(offset),
           m_layout(layout)

--- a/test/test_xstrided_view.cpp
+++ b/test/test_xstrided_view.cpp
@@ -703,6 +703,21 @@ namespace xt
         truthy = std::is_same<typename decltype(av)::shape_type, typename decltype(e)::shape_type>::value;
         EXPECT_TRUE(truthy);
     #endif
+
+        xarray<int> xa = {{1, 2, 3}, {4, 5, 6}};
+        std::vector<std::size_t> new_shape = {3, 2};
+        auto xrv = reshape_view(xa, new_shape);
+
+        if(XTENSOR_DEFAULT_LAYOUT == layout_type::row_major)
+        {
+            xarray<int> xres = {{1, 2}, {3, 4}, {5, 6}};
+            EXPECT_EQ(xrv, xres);
+        }
+        else
+        {
+            xarray<int> xres = {{1, 5}, {4, 3}, {2, 6}};
+            EXPECT_EQ(xrv, xres);
+        }
     }
 
     TEST(xstrided_view, on_xview)

